ion-schema/src/main/kotlin/com/amazon/ionschema/model/ConstraintMediator.kt

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amazon.ionschema.model
+
+/**
+ * Interactions between the Ion Schema System and [AstConstraint]s are mediated through ConstraintMediators.
+ *
+ * A ConstraintMediator can hold any kind of data associated to [ConstraintId]s, but it is intended primarily for holding
+ * things such as reader, writer, and validator functions where we need preserve some type information that would
+ * normally be lost through type-erasure.
+ */
+abstract class ConstraintMediator<R : ConstraintMediator.Registration<*>>(private val registrations: List<R>) {
```

Comment:
Having a single place to encapsulate the casts that we (the programmers) know to be safe is a good idea. I'm less keen on having that be abstract, with the presumed intention that it be sub-classed as opposed to used by composition.

Reply:
I also prefer composition to inheritance, but when I created this PR, I hadn't figured out a way to make it work in Kotlin to have a function that essentially returns `A<B>` where `A` and `B` are independent type parameters.

Thanks for challenging me on this. I returned to the problem and figured out a different way to model the types that works in Kotlin's type system.